File: projects/rtc-guard/src/rtc_guard/threat_model.py
"""Threat model for a human↔AI real-time (WebRTC) pipeline.

Structured so it can be served, rendered, and diffed — a threat model that lives in
code, not a doc that rots. Each entry maps a concrete threat on the audio/video
path to its mitigation and notes which rtc-guard control (if any) demonstrates it.
"""

THREATS: list[dict] = [
    {
        "id": "T1", "category": "token", "title": "Access-token replay",
        "threat": "A leaked join token is reused to enter a room or impersonate a "
                  "participant.",
        "mitigation": "Short TTLs + nbf; per-identity scoping; optional single-use "
                      "jti so a token can't be replayed.",
        "control": "token.mint (TTL/nbf) + verify(jti_store) — see adversary T1/T7",
    },
    {
        "id": "T2", "category": "token", "title": "Grant escalation / forgery",
        "threat": "A subscriber tampers with the grant to gain publish, or forges a "
                  "token with their own key.",
        "mitigation": "HS256 signature over the whole grant; least-privilege grant "
                      "templates; never honor alg=none.",
        "control": "token.verify signature check — see adversary T3/T4/T5",
    },
    {
        "id": "T3", "category": "room", "title": "Room hijack / cross-room replay",
        "threat": "A token minted for one room is presented to another, or an "
                  "attacker joins a room they were never scoped to.",
        "mitigation": "Room is part of the signed grant; verify against the expected "
                      "room; rooms are not guessable identifiers.",
        "control": "token.verify(expected_room) — see adversary T6",
    },
    {
        "id": "T4", "category": "media", "title": "Media eavesdropping",
        "threat": "An on-path attacker captures the audio/video stream.",
        "mitigation": "SRTP with DTLS-SRTP key exchange on every media path; "
                      "end-to-end encryption for sensitive rooms.",
        "control": "transport-level (documented); out of scope for the token service",
    },
    {
        "id": "T5", "category": "egress", "title": "Recording / egress exposure",
        "threat": "Room recordings or egress streams are written to storage that's "
                  "broader than the room's audience.",
        "mitigation": "Customer-controlled recording encryption; egress writes scoped "
                      "to least-privilege storage creds; audit who started egress.",
        "control": "policy/runbook (documented); egress requires a separate grant",
    },
    {
        "id": "T6", "category": "agent",
        "title": "Prompt injection over the data channel",
        "threat": "A participant sends data-channel messages that manipulate the AI "
                  "agent (exfiltration, jailbreak) as if they were instructions.",
        "mitigation": "Treat data-channel input as untrusted; scan/redact before the "
                      "LLM (an llm-gateway-style firewall); separate system context.",
        "control": "cross-refs the llm-gateway firewall; canPublishData is grant-gated",
    },
    {
        "id": "T7", "category": "server", "title": "SFU / signaling trust",
        "threat": "A compromised or rogue media server downgrades encryption or mirrors "
                  "streams.",
        "mitigation": "Pin trusted server identities; mutual auth on signaling; monitor "
                      "for unexpected SSRCs/participants; minimal server retention.",
        "control": "architecture (documented)",
    },
    {
        "id": "T8", "category": "dos", "title": "Join flood / resource exhaustion",
        "threat": "Mass token requests or joins exhaust room/SFU capacity.",
        "mitigation": "Rate-limit minting; per-identity caps; short TTLs limit the "
                      "blast radius of any one leaked credential.",
        "control": "mint policy (documented)",
    },
]


def threat_model() -> dict:
    by_cat: dict[str, int] = {}
    for t in THREATS:
        by_cat[t["category"]] = by_cat.get(t["category"], 0) + 1
    return {"threats": THREATS, "count": len(THREATS), "by_category": by_cat}

File: projects/relaytoken/internal/threatmodel/threatmodel.go
// Package threatmodel is the static, reviewable WebRTC / realtime-AI threat model:
// each entry maps a threat to its mitigation and the concrete control in this
// service (or the surrounding system) that implements it.
package threatmodel

// Entry is one row of the threat model.
type Entry struct {
	ID         string `json:"id"`
	Threat     string `json:"threat"`
	Vector     string `json:"vector"`
	Mitigation string `json:"mitigation"`
	Control    string `json:"control"`
}

// Model returns the full threat model.
func Model() []Entry {
	return []Entry{
		{
			ID:         "TM-1",
			Threat:     "Data-channel prompt injection into the voice-AI agent",
			Vector:     "A participant publishes crafted text on the data channel that the AI agent treats as instructions, hijacking its behavior.",
			Mitigation: "Scope canPublishData to roles that genuinely need it; treat all data-channel input as untrusted user content, never as a system prompt.",
			Control:    "Role templates default subscribers to canPublishData=false; the grant linter flags a listener with data-channel publish as a prompt-injection surface.",
		},
		{
			ID:         "TM-2",
			Threat:     "Egress / recording exposure",
			Vector:     "A token or room config carries cloud storage credentials, or recording is enabled without authorization, leaking media to attacker-controlled storage.",
			Mitigation: "Never embed storage secrets in client tokens; gate egress behind a server-side, admin-only path.",
			Control:    "Tokens carry only a VideoGrant (no egress/storage credentials); the upstream library refuses sensitive credentials in client tokens unless explicitly allowed server-side.",
		},
		{
			ID:         "TM-3",
			Threat:     "SFU / signaling trust abuse",
			Vector:     "A forged or tampered token is accepted by the media/signaling server, granting an attacker join, publish, or admin rights.",
			Mitigation: "Verify every token's HS256 signature against the server secret; never trust the payload without a valid signature; reject alg=none.",
			Control:    "Verify uses the upstream signature verifier; the adversary suite proves forged-signature, alg=none, and payload-tamper tokens are all rejected.",
		},
		{
			ID:         "TM-4",
			Threat:     "Cross-room replay",
			Vector:     "A valid token issued for one room is replayed to join a different room the holder was never authorized for.",
			Mitigation: "Bind every token to a single room scope and re-assert that scope at verify time; refuse scope-less tokens for room-scoped checks.",
			Control:    "Mint always sets a room; Verify rejects a room mismatch and refuses an empty room scope; the adversary suite covers cross-room replay.",
		},
		{
			ID:         "TM-5",
			Threat:     "Capability escalation",
			Vector:     "An attacker edits the grant payload (e.g. flips canPublish or roomAdmin) to gain privileges beyond what was issued.",
			Mitigation: "Bind capabilities into the signed claims so any payload edit invalidates the signature; verify required capability server-side.",
			Control:    "Capabilities live in the signed VideoGrant; the adversary suite flips canPublish/roomAdmin and confirms the tamper breaks verification.",
		},
		{
			ID:         "TM-6",
			Threat:     "Long-lived / stolen token replay",
			Vector:     "A leaked token with a long or absent TTL remains usable for an extended window.",
			Mitigation: "Issue short, explicit TTLs; clamp requested lifetimes to a policy ceiling; prefer minting per-session.",
			Control:    "Mint clamps TTL to [1m, 12h] with a 1h default; the grant linter flags missing or over-long TTLs.",
		},
		{
			ID:         "TM-7",
			Threat:     "Join-flood / connection DoS",
			Vector:     "An attacker rapidly joins or churns connections to exhaust SFU capacity and degrade real-time media for legitimate users.",
			Mitigation: "Rate-limit joins per identity/room at the edge; cap room participant counts; require a fresh, short-TTL token per join.",
			Control:    "Per-room, short-TTL, role-scoped tokens make each join individually authorized and revocable by expiry; rate limiting is enforced at the signaling tier (out of scope for the token service).",
		},
		{
			ID:         "TM-8",
			Threat:     "Token-mint abuse",
			Vector:     "An attacker who reaches the mint endpoint requests over-broad grants or floods minting to manufacture credentials at scale.",
			Mitigation: "Restrict minting to vetted role templates; rate-limit the mint endpoint; never accept a raw capability set from an untrusted caller.",
			Control:    "Mint accepts only a role template (publisher/subscriber/admin) and a room — not arbitrary capabilities — so a caller cannot mint a grant the template does not permit.",
		},
	}
}
